pkg/util/cmapichecker/cmapichecker.go

```diff
@@ -0,0 +1,166 @@
+	regexErrWebhookConnectionFailure              = regexp.MustCompile(`failed calling webhook "(.*)\.cert-manager\.io": Post "(.*)\/mutate(.*)": (.*): connect: connection refused$`)
+	regexErrWebhookCertificateFailure             = regexp.MustCompile(`Post "(.*)": x509: certificate signed by unknown authority`)
+)
+
+type ApiCheckError struct {
+	SimpleError     error
+	UnderlyingError error
+}
+
+func (e *ApiCheckError) Error() string {
+	return e.SimpleError.Error()
+}
+
+func (e *ApiCheckError) Cause() error {
+	return e.UnderlyingError
+}
+
+// Interface is used to check that the cert-manager CRDs have been installed and are usable.
+type Interface interface {
+	Check(context.Context) *ApiCheckError
+}
+
+type cmapiChecker struct {
+	// The client controller-runtime client.New function fails if can't reach
+	// the API server, so we load it lazily, to avoid breaking integration tests
+	// which rely on being able to start the webhook server before the API
+	// server.
+	clientBuilder func() (client.Client, error)
+
+	client client.Client
+}
+
+// New returns a cert-manager API checker
+func New(restcfg *rest.Config, scheme *runtime.Scheme, namespace string) (Interface, error) {
+	if err := cmapi.AddToScheme(scheme); err != nil {
+		return nil, errors.Wrap(err, "while configuring scheme")
+	}
+	return &cmapiChecker{
+		clientBuilder: func() (client.Client, error) {
+			cl, err := client.New(restcfg, client.Options{
+				Scheme: scheme,
+			})
+			if err != nil {
+				return nil, errors.Wrap(err, "while creating client")
+			}
+			return client.NewNamespacedClient(client.NewDryRunClient(cl), namespace), nil
+		},
+	}, nil
+}
+
+func (o *cmapiChecker) Client() (client.Client, error) {
+	if o.client != nil {
+		return o.client, nil
+	}
+
+	cl, err := o.clientBuilder()
+	if err != nil {
+		return nil, err
+	}
+	o.client = cl
+
+	return o.client, nil
+}
+
+// Check attempts to perform a dry-run create of a cert-manager *v1alpha2*
+// Certificate resource in order to verify that CRDs are installed and all the
+// required webhooks are reachable by the K8S API server.
+// We use v1alpha2 API to ensure that the API server has also connected to the
+// cert-manager conversion webhook.
+func (o *cmapiChecker) Check(ctx context.Context) *ApiCheckError {
+	cert := &cmapi.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "cmapichecker-",
+		},
+		Spec: cmapi.CertificateSpec{
+			DNSNames:   []string{"cmapichecker.example"},
+			SecretName: "cmapichecker",
+			IssuerRef: cmmeta.ObjectReference{
+				Name: "cmapichecker",
+			},
+		},
+	}
+
+	// while creating client: Get "http://localhost:8080/api?timeout=32s": dial tcp 127.0.0.1:8080: connect: connection refused
+	cl, err := o.Client()
+	if err != nil {
+		return &ApiCheckError{
+			SimpleError:     ErrAPIServerUnreachable,
+			UnderlyingError: err,
+		}
```

Comment:
Are there any other reasons that `Client()` might fail, reasons other than server unreachable?

Reply:
I don't think so.